Add Deploy to Heroku button

spapas opened the pull request:
This adds a deploy to heroku button (https://devcenter.heroku.com/articles/heroku-button) to Wagtaildemo.

It can be tested right now by clicking the button (at spapas/wagtaildemo), but it will use my fork instead of torchbox/wagtaildemo. After merging we need to change the ``README.md`` and ``app.json`` to point to the correct repository.

Finally, because of how Heroku works, the filesystem is read only so the uploaded images that would be copied through ``load_initial_data`` to the ``MEDIA`` folder will not be copied.

tomdyson commented:
This is great, @spapas! I've just tried it successfully. 

What do you think is the best solution for image uploads? An S3 storage backend, with the bucket and access key details set in the Heroku startup process?

spapas commented:
@tomdyson probably that's the best solution to be able to use user uploaded files in wagtaildemo in Heroku. However this has the disadvantage that users that don't have S3 won't be able to use the "Deploy to Heroku button" (I don't have S3 and from what I've seen you need to give your credit card details in order to get one so many people won't do it).

To resolve the above maybe we need to configure settings/heroku.py to check if the user has actually set S3 credentials and if not fall back to the simple file storage (without S3).

kaedroho commented on requirements.txt:
```diff
@@ -0,0 +1 @@
+-r requirements/heroku.txt
```
My personal preference here would be to move all requirements (including dev and heroku) into a single ``requirements.txt`` (I'm not much of a fan of splitting up the requirements files). I think this would really improve readability and work well with Heroku too.

spapas replied:
Yes, we could do this. However we'd need to merge all requirements files in a single wagtaildemo/requirements.txt (Heroku needs the requirements.txt to be on the project's root directory).

kaedroho replied:
Yep, I think that's a good idea. That's how it's done in the ``wagtail-template`` repo: https://github.com/torchbox/wagtail-template/blob/master/requirements.txt

gasman replied:
As per [my email to wagtail-dev last July](https://gist.github.com/gasman/0e6b5551104830e51e5f) (which I really should move forward with at some point...) I'm broadly in favour of abolishing the requirements/dev.txt / requirements/production.txt separation. However, wagtaildemo is a slightly special case, in that the priority is to make it work on lots of different setups, rather than to ensure that dev and production environments are consistent.

Practically, I think the only issue would be if the Heroku or production requirements involved a package that was awkward to install (e.g. requiring external C libraries) - in which case we shouldn't inflict that on all users who just want a simple local setup. If not, then this is just a theoretical concern, and I'd be happy to go ahead with merging the requirements files.

tomdyson commented:
@spapas you're right, we shouldn't require S3 for Heroku users. I like your suggestion ('check if the user has actually set S3 credentials and if not fall back to the simple file storage'). But what I don't understand is how I'm already able to upload images to the newly created instance on Heroku? Where are they being stored? If I can do this, why can't ``load_initial_data``?

kaedroho commented:
@tomdyson 
Each Dyno on Heroku has its own read-write filesystem that is based on your apps image. When you run comands like ``load_initial_data`` Heroku will fire up a fresh "one-off" Dyno with a new filesystem, run a command then delete the dyno (this is very fast as a Dyno is just a process in a container, not a VM).

See: https://devcenter.heroku.com/articles/one-off-dynos

When you upload through the web interface, you are uploading to the same Dyno that's serves the media so it is able to find your images. This isn't good practise though as if you decide to add more Dynos, the new ones won't have access to your media and your media may get accidentally deleted when you scale down.

tomdyson commented:
@kaedroho thanks for explaining. Can we do:

- if the user supplies S3 details, use an S3 storage backend
- otherwise, accept the bad-practice-same-Dyno approach, and add the images to the ``MEDIA`` folder as a final step

?

tomdyson commented:
And can / should we set a default username and password?

kaedroho commented:
The bad-practise way should be possible by sneaking in a copy command into the Procfile.

Something like:
```
web: cp -R <source_dir> <dest_dir> && gunicorn wagtaildemo.heroku_wsgi
```

This will tell the dyno to copy the media files into the media directory when it first boots up. I've not tried it with Heroku before, but it works with Docker which works almost the same way.

tomdyson commented:
@kaedroho can we do this only when they don't supply S3 credentials?

@spapas please go ahead with these changes, along with creating a default admin user (admin / changeme) and updating the README to reflect this (i.e. removing ``./manage.py createsuperuser``).

kaedroho commented:
@tomdyson 
Yes, we can. We would need to make a bash script which detects whether the environment variable is set and copies the files over if not.

spapas commented:
Added various changes based on the discussion here. Images seem to be working now.

Sorry for the long list of minor changes but Heroku Button will download the master branch of your repository. Should I rebase+squash these commits? I didn't do it because I wasn't sure if the PR will remain open after that. Can somebody confirm that a ``push --force`` with a different history won't close the PR?

Two more things:

1. The required env variables + checks have been added however I wasn't able to test that AWS S3 deployment actually works (no AWS credentials)

2. @kaedroho instead of adding the check you mention to copy the files, I just created a heroku.sh file that runs ``migrate`` and ``load_initial_data`` every time before starting gunicorn (using Procfile) and it is working fine (``load_initial_data`` copies the images without problem).

3. I didnt' merge the requirements in a single file since that could be part of a differnet PR.

kaedroho commented:
Hi @spapas. The changes you have made look good to me. I think migrate belongs in "postdeploy" though so it only runs on first deployment/code updates instead of every time a dyno starts. Since we're not supporting people running multiple dynos, this is probably not a problem.

(3) I agree with that, best to keep PRs as clean as possible

kaedroho commented on demo/management/commands/load_initial_data.py:
```diff
@@ -19,3 +19,5 @@ def handle_noargs(self, **options):
         for filename in os.listdir(image_src_dir):
             shutil.copy(os.path.join(image_src_dir, filename), image_dest_dir)
+
+        
```
Minor coding style issue slipped through the net here.

tomdyson commented:
Thanks, @spapas. Images work when AWS credentials are not supplied, but not when they are - neither with ``load_initial_data`` nor by upload through the admin UI.

I can share some AWS creds with you separately, but perhaps we should forget about S3 support for now; this is an excellent demonstration and I don't think anyone will be starting from the Wagtaildemo button if they want to use Wagtail on Heroku in production.

spapas commented:
@kaedroho I've squashed the minor commits and removed S3 stuff as per @tomdyson 's suggestion. 

I didn't put migrate in postdeploy since, if there are no migrations nothing will happen so it can be left there.

kaedroho commented:
Don't forget that people copy things from wagtaildemo into their production sites though (happens in Torchbox very often).

spapas commented:
Ok I put migrate in postdeploy (I also removed the DJANGO_SETTINGS_MODULE env var since it couldn't actually be changed by users trying wagtaildemo)

kaedroho commented:
This works well with Dokku as well, with no code changes! http://wagtaildemo.kaed.uk/
